README.md
[![License](https://img.shields.io/github/license/materialsvirtuallab/megnet)]()
[![Build Status](https://travis-ci.org/materialsvirtuallab/megnet.svg?branch=master)](https://travis-ci.org/materialsvirtuallab/megnet)
[![Coverage Status](https://coveralls.io/repos/github/materialsvirtuallab/megnet/badge.svg?branch=master)](https://coveralls.io/github/materialsvirtuallab/megnet?branch=master&service=github)
[![Downloads](https://pepy.tech/badge/megnet)](https://pepy.tech/project/megnet)
[![Linting](https://github.com/materialsvirtuallab/megnet/workflows/Linting/badge.svg)](https://github.com/materialsvirtuallab/megnet/workflows/Linting/badge.svg)
[![Testing](https://github.com/materialsvirtuallab/megnet/workflows/Testing%20-%20main/badge.svg)](https://github.com/materialsvirtuallab/megnet/workflows/Testing%20-%20main/badge.svg)


[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/materialsvirtuallab/megnet/master)

# Table of Contents
* [Introduction](#introduction)
* [MEGNet Framework](#megnet-framework)
* [Installation](#installation)
* [Usage](#usage)
* [Datasets](#datasets)
* [Implementation details](#implementation-details)
* [Computing requirements](#computing-requirements)
* [Known limitations](#limitations)
* [Contributors](#contributors)
* [References](#references)

<a name="introduction"></a>
# Introduction

This repository represents the efforts of the [Materials Virtual Lab](http://www.materialsvirtuallab.org)
in developing graph networks for machine learning in materials science. It is a
work in progress and the models we have developed thus far are only based on
our best efforts. We welcome efforts by anyone to build and test models using
our code and data, all of which are publicly available. Any comments or
suggestions are also welcome (please post on the Github Issues page.)

A web app using our pre-trained MEGNet models for property prediction in
crystals is available at [http://megnet.crystals.ai](http://megnet.crystals.ai). For tutorials, please visit `notebooks` in this repo. We have also established an online simulation tool and a tutorial lecture at nanoHUB ([https://nanohub.org/resources/megnet](https://nanohub.org/resources/megnet)).

Note: A [DGL implementation of MEGNet](https://github.com/materialsvirtuallab/m3gnet-dgl) is now available. For users
trying to build their own MEGNet models, it is highly recommended you check this version out, which may be easier to
work with and extend in future.

<a name="megnet-framework"></a>
# MEGNet framework

The MatErials Graph Network (MEGNet) is an implementation of DeepMind's graph
networks[1] for universal machine learning in materials science. We have
demonstrated its success in achieving very low prediction errors in a broad
array of properties in both molecules and crystals (see
["Graph Networks as a Universal Machine Learning Framework for Molecules and Crystals"](https://doi.org/10.1021/acs.chemmater.9b01294)[2]). New releases have included our recent work on multi-fidelity materials property modeling (See ["Learning properties of ordered and disordered materials from multi-fidelity data"](https://www.nature.com/articles/s43588-020-00002-x)[3]).

Briefly, Figure 1 shows the sequential update steps of the graph network,
whereby bonds, atoms, and global state attributes are updated using information
from each other, generating an output graph.

![GraphModel diagram](resources/model_diagram_small.jpg)
<div align='center'><strong>Figure 1. The graph network update function.</strong></div>

Figure 2 shows the overall schematic of the MEGNet. Each graph network module
is preceded by two multi-layer perceptrons (known as Dense layers in Keras
terminology), constituting a MEGNet block. Multiple MEGNet blocks can be
stacked, allowing for information flow across greater spatial distances. The
number of blocks required depend on the range of interactions necessary to
predict a target property. In the final step, a `set2set` is used to map the
output to a scalar/vector property.

![GraphModel architecture](resources/model_arch_small.jpg)
<div align='center'><strong>Figure 2. Schematic of MatErials Graph Network.</strong></div>

<a name="installation"></a>
# Installation

Megnet can be installed via pip for the latest stable version:

```bash
pip install megnet
```

For the latest dev version, please clone this repo and install using:

```bash
python setup.py develop
```

<a name="usage"></a>
# Usage

Our current implementation supports a variety of use cases for users with
different requirements and experience with deep learning. Please also visit
the [notebooks directory](notebooks) for Jupyter notebooks with more detailed code examples.

## Using pre-built models

In our work, we have already built MEGNet models for the QM9 data set and
Materials Project dataset. These models are provided as serialized HDF5+JSON
files. Users who are purely interested in using these models for prediction
can quickly load and use them via the convenient `MEGNetModel.from_file`
method. These models are available in the `mvl_models` folder of this repo.
The following models are available:

* QM9 molecule data:
    - HOMO: Highest occupied molecular orbital energy
    - LUMO: Lowest unoccupied molecular orbital energy
    - Gap: energy gap
    - ZPVE: zero point vibrational energy
    - µ: dipole moment
    - α: isotropic polarizability
    - \<R2\>: electronic spatial extent
    - U0: internal energy at 0 K
    - U: internal energy at 298 K
    - H: enthalpy at 298 K
    - G: Gibbs free energy at 298 K
    - Cv: heat capacity at 298 K
    - ω1: highest vibrational frequency.
* Materials Project data:
    - Formation energy from the elements
    - Band gap
    - Log 10 of Bulk Modulus (K)
    - Log 10 of Shear Modulus (G)

The MAEs on the various models are given below:

### Performance of QM9 MEGNet-Simple models

| Property | Units      | MAE   |
|----------|------------|-------|
| HOMO     | eV         | 0.043 |
| LUMO     | eV         | 0.044 |
| Gap      | eV         | 0.066 |
| ZPVE     | meV        | 1.43  |
| µ        | Debye      | 0.05  |
| α        | Bohr^3     | 0.081 |
| \<R2\>   | Bohr^2     | 0.302 |
| U0       | eV         | 0.012 |
| U        | eV         | 0.013 |
| H        | eV         | 0.012 |
| G        | eV         | 0.012 |
| Cv       | cal/(molK) | 0.029 |
| ω1       | cm^-1      | 1.18  |

### Performance of MP-2018.6.1

| Property | Units      | MAE   |
|----------|------------|-------|
| Ef       | eV/atom    | 0.028 |
| Eg       | eV         | 0.33  |
| K_VRH    | log10(GPa) | 0.050 |
| G_VRH    | log10(GPa) | 0.079 |

### Performance of MP-2019.4.1

| Property | Units      | MAE   |
|----------|------------|-------|
| Ef       | eV/atom    | 0.026 |
| Efermi   | eV         | 0.288 |

New models will be added as they are developed in the [mvl_models](mvl_models)
folder. Each folder contains a summary of model details and benchmarks. For
the initial models and bencharmks comparison to previous models,
please refer to ["Graph Networks as a Universal Machine Learning Framework for Molecules and Crystals"](https://doi.org/10.1021/acs.chemmater.9b01294)[2].

Below is an example of crystal model usage:

```python
from megnet.utils.models import load_model
from pymatgen.core import Structure, Lattice

# load a model in megnet.utils.models.AVAILABLE_MODELS
model = load_model("logK_MP_2018")

# We can construct a structure using pymatgen
structure = Structure(Lattice.cubic(3.167),
            ['Mo', 'Mo'], [[0, 0, 0], [0.5, 0.5, 0.5]])


# Use the model to predict bulk modulus K. Note that the model is trained on
# log10 K. So a conversion is necessary.
predicted_K = 10 ** model.predict_structure(structure).ravel()[0]
print(f'The predicted K for {structure.composition.reduced_formula} is {predicted_K:.0f} GPa.')
```
A full example is in [notebooks/crystal_example.ipynb](notebooks/crystal_example.ipynb).

For molecular models, we have an example in
[notebooks/qm9_pretrained.ipynb](notebooks/qm9_pretrained.ipynb).
We support prediction directly from a pymatgen molecule object. With a few more
lines of code, the model can predict from `SMILES` representation of molecules,
as shown in the example. It is also straightforward to load a `xyz` molecule
file with pymatgen and predict the properties using the models. However, the
users are generally not advised to use the `qm9` molecule models for other
molecules outside the `qm9` datasets, since the training data coverage is
limited.

Below is an example of predicting the "HOMO" of a smiles representation

```python
from megnet.utils.molecule import get_pmg_mol_from_smiles
from megnet.models import MEGNetModel

# same model API for molecule and crystals, you can also use the load_model method
# as in previous example
model = MEGNetModel.from_file('mvl_models/qm9-2018.6.1/HOMO.hdf5')
# Need to convert SMILES into pymatgen Molecule
mol = get_pmg_mol_from_smiles("C")
model.predict_structure(mol)
```

## Training a new MEGNetModel from structures

For users who wish to build a new model from a set of crystal structures with
corresponding properties, there is a convenient `MEGNetModel` class for setting
up and training the model. By default, the number of MEGNet blocks is 3 and the
atomic number Z is used as the only node feature (with embedding).

```python
from megnet.models import MEGNetModel
from megnet.data.crystal import CrystalGraph
import numpy as np

nfeat_bond = 10
r_cutoff = 5
gaussian_centers = np.linspace(0, r_cutoff + 1, nfeat_bond)
gaussian_width = 0.5
graph_converter = CrystalGraph(cutoff=r_cutoff)
model = MEGNetModel(graph_converter=graph_converter, centers=gaussian_centers, width=gaussian_width)

# Model training
# Here, `structures` is a list of pymatgen Structure objects.
# `targets` is a corresponding list of properties.
model.train(structures, targets, epochs=10)

# Predict the property of a new structure
pred_target = model.predict_structure(new_structure)
```
Note that for realistic models, the `nfeat_bond` can be set to 100 and `epochs` can be 1000.
In some cases, some structures within the training pool may not be valid (containing isolated atoms),
then one needs to use `train_from_graphs` method by training only on the valid graphs.

Following the previous example,
```python
model = MEGNetModel(graph_converter=graph_converter, centers=gaussian_centers, width=gaussian_width)
graphs_valid = []
targets_valid = []
structures_invalid = []
for s, p in zip(structures, targets):
    try:
        graph = model.graph_converter.convert(s)
        graphs_valid.append(graph)
        targets_valid.append(p)
    except:
        structures_invalid.append(s)

# train the model using valid graphs and targets
model.train_from_graphs(graphs_valid, targets_valid)
```
For model details and benchmarks, please refer to ["Graph Networks as a Universal Machine Learning Framework for Molecules and Crystals"](https://doi.org/10.1021/acs.chemmater.9b01294)[2]


### Training multi-fidelity graph networks

Please see the folder `multifidelity` for specific examples.

### Pre-trained elemental embeddings

A key finding of our work is that element embeddings from trained formation
energy models encode useful chemical information that can be transferred
learned to develop models with smaller datasets (e.g. elastic constants,
band gaps), with better converegence and lower errors. These embeddings are
also potentially useful in developing other ML models and applications. These
embeddings have been made available via the following code:

```python
from megnet.data.crystal import get_elemental_embeddings

el_embeddings = get_elemental_embeddings()
```

An example of transfer learning using the elemental embedding from formation
energy to other models, please check [notebooks/transfer_learning.ipynb](notebooks/transfer_learning.ipynb).

## Customized Graph Network Models

For users who are familiar with deep learning and Keras and wish to build
customized graph network based models, the following example outlines how a
custom model can be constructed from `MEGNetLayer`, which is essentially our
implementation of the graph network using neural networks:

```python
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.models import Model
from megnet.layers import MEGNetLayer, Set2Set

n_atom_feature= 20
n_bond_feature = 10
n_global_feature = 2

# Define model inputs
int32 = 'int32'
x1 = Input(shape=(None, n_atom_feature)) # atom feature placeholder
x2 = Input(shape=(None, n_bond_feature)) # bond feature placeholder
x3 = Input(shape=(None, n_global_feature)) # global feature placeholder
x4 = Input(shape=(None,), dtype=int32) # bond index1 placeholder
x5 = Input(shape=(None,), dtype=int32) # bond index2 placeholder
x6 = Input(shape=(None,), dtype=int32) # atom_ind placeholder
x7 = Input(shape=(None,), dtype=int32) # bond_ind placeholder
xs = [x1, x2, x3, x4, x5, x6, x7]

# Pass the inputs to the MEGNetLayer layer
# Here the list are the hidden units + the output unit,
# you can have others like [n1] or [n1, n2, n3 ...] if you want.
out = MEGNetLayer([32, 16], [32, 16], [32, 16], pool_method='mean', activation='relu')(xs)

# the output is a tuple of new graphs V, E and u
# Since u is a per-structure quantity,
# we can directly use it to predict per-structure property
out = Dense(1)(out[2])

# Set up the model and compile it!
model = Model(inputs=xs, outputs=out)
model.compile(loss='mse', optimizer='adam')
```

With less than 20 lines of code, you have built a graph network model that is
ready for materials property prediction!

<a name="implementation-details"></a>
# Implementation details

Graph networks[1] are a superclass of graph-based neural networks. There are a
few innovations compared to conventional graph-based neural neworks.

* Global state attributes are added to the node/edge graph representation.
  These features work as a portal for structure-independent features such as
  temperature, pressure etc and also are an information exchange placeholder
  that facilitates information passing across longer spatial domains.
* The update function involves the message interchange among all three levels
  of information, i.e., the node, bond and state information. It is therefore a
  highly general model.

The `MEGNet` model implements two major components: (a) the `graph network`
layer and (b) the `set2set` layer.[4] The layers are based on
[keras](https://keras.io/) API and is thus compatible with other keras modules.

Different crystals/molecules have different number of atoms. Therefore it is
impossible to use data batches without padding the structures to make them
uniform in atom number. `MEGNet` takes a different approach. Instead of making
structure batches, we assemble many structures into one giant structure, and
this structure has a vector output with each entry being the target value for
the corresponding structure. Therefore, the batch number is always 1.

Assuming a structure has N atoms and M bonds, a structure graph is represented
as **V** (nodes/vertices, representing atoms), **E** (edges, representing bonds)
and **u** (global state vector). **V** is a N\*Nv matrix. **E** comprises of a
M\*Nm matrix for the bond attributes and index pairs (rk, sk) for atoms
connected by each bond. **u** is a vector with length Nu. We vectorize rk and
sk to form `index1` and `index2`, both are vectors with length M. In summary,
the graph is a data structure with **V** (N\*Nv), **E** (M\*Nm), **u**
(Nu, ), `index1` (M, ) and `index2` (M, ).

We then assemble several structures together. For **V**, we directly append the
atomic attributes from all structures, forming a matrix (1\*N'\*Nv), where
N' > N. To indicate the belongingness of each atom attribute vector, we use a
`atom_ind` vector. For example if `N'=5` and the first 3 atoms belongs to the
first structure and the remainings the second structure, our `atom_ind` vector
would be `[0, 0, 0, 1, 1]`. For the bond attribute, we perform the same
appending method, and use `bond_ind` vector to indicate the bond belongingness.
For `index1` and `index2`, we need to shift the integer values. For example,
if `index1` and `index2` are `[0, 0, 1, 1]` and `[1, 1, 0, 0]` for structure 1
and are `[0, 0, 1, 1]` and `[1, 1, 0, 0]` for structure two. The assembled
indices are `[0, 0, 1, 1, 2, 2, 3, 3]` and `[1, 1, 0, 0, 3, 3, 2, 2]`. Finally,
**u** expands a new dimension to take into account of the number of structures,
and becomes a 1\*Ng\*Nu tensor, where Ng is the number of structures. `1` is
added as the first dimension of all inputs because we fixed the batch size to
be 1 (1 giant graph) to comply the keras inputs requirements.

In summary the inputs for the model is **V** (1\*N'\*Nv), **E** (1\*M'\*Nm),
**u** (1\*Ng\*Nu), `index1` (1\*M'), `index2` (1\*M'), `atom_ind` (1\*N'), and
`bond_ind` (1\*M'). For Z-only atomic features, **V** is a (1\*N') vector.

<a name="datasets"></a>
# Data sets

To aid others in reproducing (and improving on) our results, we have provided
our MP-crystals-2018.6.1 crystal data set via [figshare](https://figshare.com/articles/Graphs_of_materials_project/7451351)[5].
The MP-crystals-2018.6.1 data set comprises the DFT-computed energies and
band gaps of 69,640 crystals from the [Materials Project](http://www.materialsproject.org)
obtained via the [Python Materials Genomics (pymatgen)](http://pymatgen.org)
interface to the Materials Application Programming Interface (API)[6] on
June 1, 2018. The crystal graphs were constructed using a radius cut-oﬀ of 4
angstroms. Using this cut-oﬀ, 69,239 crystals do not form isolated atoms and
are used in the models. A subset of 5,830 structures have elasticity data that
do not have calculation warnings and will be used for elasticity models.

The molecule data set used in this work is the QM9 data set 30 processed by
Faber et al.[7] It contains the B3LYP/6-31G(2df,p)-level DFT calculation
results on 130,462 small organic molecules containing up to 9 heavy atoms.

<a name="computing-requirements"></a>
# Computing requirements

Training: It should be noted that training MEGNet models, like other deep
learning models, is fairly computationally intensive with large datasets. In
our work, we use dedicated GPU resources to train MEGNet models with 100,000
crystals/molecules. It is recommended that you do the same.

Prediction: Once trained, prediction using MEGNet models are fairly cheap.
For example, the http://megnet.crystals.ai web app runs on a single hobby dyno
on Heroku and provides the prediction for any crystal within seconds.

<a name="limitations"></a>
# Known limitations

- `isolated atoms` error. This error occurs when using the given cutoff in the model (4A for
2018 models and 5A for 2019 models), the crystal structure contains isolated atoms, i.e.,
no neighboring atoms are within the distance of `cutoff`. Most of the time, we can just
discard the structure, since we found that those structures tend to have a high energy above
hull (less stable). If you think this error is an essential issue for a particular problem,
please feel free to email us and we will consider releasing a new model with increased cutoff.

<a name="contributors"></a>
# Contributors
1. Chi Chen from the Materials Virtual Lab is the lead developer of MEGNet.
2. Shyue Ping Ong and other members of the Materials Virtual Lab contributes to general improvements
   of MEGNet and its applications.
3. Logan Ward has made extensive contributions, especially to the development of molecular graph
   portions of MEGNet.

<a name="references"></a>
# References

1. Battaglia, P. W.; Hamrick, J. B.; Bapst, V.; Sanchez-Gonzalez, A.;

CHANGES.md
# v1.3.2
* Disable progress bar for readability.

# v1.3.1
* Disable notifications.

# v1.3.0
* Enable isolated atom calculations

# v1.2.9
* Security fix of tensorflow
* Docstring fix @a-ws-m
* Minor format fixes

# v1.2.8
* Revert to use tf.gather with warnings for speed concerns

# v1.2.7
* Add swish activation function
* Remove training warning
* Update tutorial

# v1.2.6
* Stability improvements

# v1.2.5
* Include data files in wheel

# v1.2.4
* Add pre-trained multi-fidelity models
* Add utils for disordered site representation

# v1.2.3
* Add more multi-fidelity plots

# v1.2.2
* Revising plotting of multifidelity work

# v1.2.1
* Release for minting DOI

# v1.2.0
* Bug fix

# v1.1.9
* Add multi-fidelity model examples
* Add sample weights for model training
* Add default optimizer gradient norm clip

# v1.1.8
* Bug fix of megnet descriptors

# v1.1.7
* Update the model training mechanism and move Gaussian expansion to tensorflow, training speed up 100%

# v1.1.6
* minor fix to include more linear readout option
* add data type control
* refactor local_env

# v1.1.5
* Code refactor and reformat
* Fix tensorflow and numpy type compatibility issues

# v1.1.4
* Update to tensorflow.keras API instead of using keras

# v1.1.3
* Download mvl_models from figshare if not present

# v1.1.2
* Add mvl_models in wheel release file

# v1.1.0
* Bug fix and version correction

# v1.0.3
* Fix bug brought by migrating to tensorflow 2.0
* New elasticity models trained on 2019 MP data base
* Add meg command line tools

# v1.0.2
* Add mypy typing hint for non-tensorflow codes
* Update keras to 2.3.1 to fix thread-safety issues

# v1.0.1
* New find_points_in_spheres algorithm in pymatgen for graph construction

# v1.0.0
* Tensorflow 2.0 version.

# v0.3.4
* Change `convertor` to `converter` in all model APIs
* Improve `ReduceLRUponNan` callback function
* @WardLT major contributions to the `MolecularGraph` class
* Add serialization methods for `local_env` classes
* delete `data/mp.py`

# v0.3.3
* GraphModel and MEGNetModel now supports a metadata tag, which is included in
  the JSON. (suggestion of @mhorton).
* Misc bug fixes for edge cases as well as improved error messages for
  mismatches in inputs.

# v0.3.2
* Implement the option for a scaler in models, which is used in efermi models at
  the moment but also can be helpful for extensive quantities.

# v0.3.1
* Minor fixes to setup.py and licenses.

# v0.3.0
* Proper fix to release on PyPi.

# v0.2.0
* Major refactoring to conform to OOP principles. Note that the
  changes are not backwards compatible, but many things are a lot
  simpler. We do not expect much disruption to existing users.
* Added pre-trained models developed in our work for users who
  wish to simply use them for prediction.
* Major improvements to README and documentation.

# v0.1.0

* Bug fix for dimension problem when only one atom in structure

# v0.0.1

* Initial release


megnet/model.py
from keras.optimizers import Adam
from keras.layers import Dense, Input, Concatenate, Add, Embedding, Dropout
from megnet.layers import MEGNet, Set2Set
from megnet.activations import softplus2
from megnet.losses import mse_scale
from keras.regularizers import l2

from keras.models import Model as KerasModel
from megnet.callbacks import ModelCheckpointMAE, ManualStop
from megnet.utils.general_utils import expand_1st
from megnet.data.graph import GraphBatchDistanceConvert, GraphBatchGenerator
from megnet.data.crystal import graphs2inputs
import numpy as np
from sklearn.preprocessing import StandardScaler
import os

from monty.serialization import dumpfn, loadfn
from monty.json import MontyEncoder, MontyDecoder

pjoin = os.path.join


class GraphModel:
    """
    Wrapper of keras.Model.
    We add methods to train the model from (structures, targets) pairs

    Args:
        model: (keras model)
        graph_convertor: (object) a object that turns a structure to a graph
        distance_convertor: (object) expand the distance value to a vector

    Or one can specify the maximum radius (max_r), number of basis (n_basis) and
    the Gaussian width (width) for constructing the distance convertor
    """

    def __init__(self,
                 model,
                 graph_convertor,
                 distance_convertor):
        self.model = model
        self.graph_convertor = graph_convertor
        self.distance_convertor = distance_convertor
        self.yscaler = StandardScaler()

    def __getattr__(self, p):
        return getattr(self.model, p)

    def train(self,
              train_structures,
              train_targets,
              validation_structures=None,
              validation_targets=None,
              epochs=1000,
              batch_size=128,
              verbose=1,
              callbacks=None,
              prev_model=None,
              **kwargs):
        """
        :param train_structures: (list) list of pymatgen structures
        :param train_targets: (list) list of target values
        :param validation_structures: (list) list of pymatgen structures as
            validation
        :param validation_targets: (list) list of validation targets
        :param epochs: (int) number of epochs
        :param batch_size: (int) training batch size
        :param verbose: (int) keras fit verbose, 0 no progress bar, 1 only at
            the epoch end and 2 every batch
        :param callbacks: (list) megnet or keras callback functions for training
        :param prev_model: (str) file name for previously saved model
        :param kwargs:
        :return:
        """
        train_graphs = [self.graph_convertor(i) for i in train_structures]
        if validation_structures is not None:
            val_graphs = [self.graph_convertor(i) for i in validation_structures]
        else:
            val_graphs = None

        self.train_from_graphs(train_graphs,
                               train_targets,
                               validation_graphs=val_graphs,
                               validation_targets=validation_targets,
                               epochs=epochs,
                               batch_size=batch_size,
                               verbose=verbose,
                               callbacks=callbacks,
                               prev_model=prev_model,
                               **kwargs
                               )

    def train_from_graphs(self,
                          train_graphs,
                          train_targets,
                          validation_graphs=None,
                          validation_targets=None,
                          epochs=1000,
                          batch_size=128,
                          verbose=1,
                          callbacks=None,
                          prev_model=None,
                          **kwargs
                          ):

        # load from saved model
        if prev_model:
            self.load_weights(prev_model)
        dirname = kwargs.get('dirname', 'callback')
        if not os.path.isdir(dirname):
            os.makedirs(dirname)
        if callbacks is None:
            callbacks = [ManualStop()]
        train_targets = self.yscaler.fit_transform(np.array(train_targets).reshape((-1, 1))).ravel()
        if validation_graphs is not None:
            filepath = pjoin(dirname, 'val_mae_{epoch:05d}_{val_mae:.6f}.hdf5')
            validation_targets = self.yscaler.transform(np.array(validation_targets).reshape((-1, 1))).ravel()
            val_inputs = graphs2inputs(validation_graphs, validation_targets)

            val_generator = self._create_generator(*val_inputs,
                                                   batch_size=batch_size)
            steps_per_val = int(np.ceil(len(validation_graphs) / batch_size))
            # print(steps_per_val)
            callbacks.extend([ModelCheckpointMAE(filepath=filepath,
                                                 save_best_only=True,
                                                 save_weights_only=False,
                                                 val_gen=val_generator,
                                                 steps_per_val=steps_per_val,
                                                 y_scaler=self.yscaler)])
        else:
            val_generator = None
            steps_per_val = None
        train_inputs = graphs2inputs(train_graphs, train_targets)
        train_generator = self._create_generator(*train_inputs, batch_size=batch_size)
        steps_per_train = int(np.ceil(len(train_graphs) / batch_size))
        self.fit_generator(train_generator, steps_per_epoch=steps_per_train,
                           validation_data=val_generator, validation_steps=steps_per_val,
                           epochs=epochs, verbose=verbose, callbacks=callbacks)

    def predict_structure(self, structure):
        """
        Predict property from structure
        :param structure:
        :return:
        """
        graph = self.graph_convertor(structure)
        gnode = [0] * len(graph['node'])
        gbond = [0] * len(graph['index1'])
        inp = [expand_1st(graph['node']),
               expand_1st(self.distance_convertor.convert(graph['distance'])),
               expand_1st(np.array(graph['state'])),
               expand_1st(np.array(graph['index1'])),
               expand_1st(np.array(graph['index2'])),
               expand_1st(np.array(gnode)),
               expand_1st(np.array(gbond)),
               ]
        return self.yscaler.inverse_transform(self.predict(inp).reshape((-1, 1))).ravel()

    def _create_generator(self, *args, **kwargs):
        if self.distance_convertor is not None:
            kwargs.update({'distance_convertor': self.distance_convertor})
            return GraphBatchDistanceConvert(*args, **kwargs)
        else:
            return GraphBatchGenerator(*args, **kwargs)

    def save_model(self, filename):
        """
        Save the model to a keras model hdf5 and a json config for additional
        convertors

        :param filename: (str)
        :return:
        """
        self.model.save(filename)
        dumpfn(
            {
                'distance_convertor': self.distance_convertor,
                'graph_convertor': self.graph_convertor
            },
            filename + '.json'
        )

    @classmethod
    def from_file(cls, filename):
        """
        Class method to load model from
            filename for keras model
            filename.json for additional convertors

        :param filename: (str)
        :return: Model
        """
        configs = loadfn(filename + '.json')
        model = load_megnet_model(filename)
        configs.update({'model': model})
        return cls(**configs)


def megnet_model(n_connect,
                 n_global,
                 n_feature=None,
                 n_blocks=3,
                 lr=1e-3,
                 n1=64,
                 n2=32,
                 n3=16,
                 n_vocal=95,
                 embedding_dim=16,
                 n_pass=3,
                 n_target=1,
                 act=softplus2,
                 is_classification=False,
                 loss="mse",
                 l2_coef=None,
                 dropout=None,
                 graph_convertor=None,
                 distance_convertor=None):
    """
    construct a graph network model with or without explicit atom features
    if n_feature is specified then a general graph model is assumed, otherwise
    a crystal graph model with z number as
    atom feature is assumed.

    :param n_connect: (int) number of bond features
    :param n_global: (int) number of state features
    :param n_feature: (int) number of atom features
    :param n_blocks: (int) number of MEGNet block
    :param lr: (float) learning rate
    :param n1: (int) number of hidden units in layer 1 in MEGNet
    :param n2: (int) number of hidden units in layer 2 in MEGNet
    :param n3: (int) number of hidden units in layer 3 in MEGNet
    :param n_vocal: (int) number of total element
    :param embedding_dim: (int) number of embedding dimension
    :param n_pass: (int) number of recurrent steps in Set2Set layer
    :param n_target: (int) number of output targets
    :param act: (object) activation function
    :param l2_coef: (float or None) l2 regularization parameter
    :param is_classification: (bool) whether it is a classifiation task
    :param loss: (object or str) loss function
    :param dropout: (float) dropout rate
    :param graph_convertor: (object) object that exposes a "convert" method for structure to graph conversion
    :param distance_convertor: (object) object that exposes a "convert" method for distance to expanded vector conversion
    :return: keras model object
    """
    int32 = 'int32'
    if n_feature is None:
        x1 = Input(shape=(None,), dtype=int32)  # only z as feature
        x1_ = Embedding(n_vocal, embedding_dim)(x1)
    else:
        x1 = Input(shape=(None, n_feature))
        x1_ = x1
    x2 = Input(shape=(None, n_connect))
    x3 = Input(shape=(None, n_global))
    x4 = Input(shape=(None,), dtype=int32)
    x5 = Input(shape=(None,), dtype=int32)
    x6 = Input(shape=(None,), dtype=int32)
    x7 = Input(shape=(None,), dtype=int32)

    if l2_coef is not None:
        reg = l2(l2_coef)
    else:
        reg = None

    # two feedforward layers
    def ff(x, n_hiddens=[n1, n2]):
        out = x
        for i in n_hiddens:
            out = Dense(i, activation=act, kernel_regularizer=reg)(out)
        return out

    # a block corresponds to two feedforward layers + one MEGNet layer
    # Note the first block does not contain the feedforward layer since
    # it will be explicitly added before the block
    def one_block(a, b, c, has_ff=True):
        if has_ff:
            x1_ = ff(a)
            x2_ = ff(b)
            x3_ = ff(c)
        else:
            x1_ = a
            x2_ = b
            x3_ = c
        out = MEGNet([n1, n1, n2], [n1, n1, n2], [n1, n1, n2], pool_method='mean', activation=act)(
            [x1_, x2_, x3_, x4, x5, x6, x7])

        x1_temp = out[0]
        x2_temp = out[1]
        x3_temp = out[2]
        if dropout:
            x1_temp = Dropout(dropout)(x1_temp)
            x2_temp = Dropout(dropout)(x2_temp)
            x3_temp = Dropout(dropout)(x3_temp)
        return x1_temp, x2_temp, x3_temp

    x1_ = ff(x1_)
    x2_ = ff(x2)
    x3_ = ff(x3)
    for i in range(n_blocks):
        if i == 0:
            has_ff = False
        else:
            has_ff = True
        x1_1 = x1_
        x2_1 = x2_
        x3_1 = x3_
        x1_1, x2_1, x3_1 = one_block(x1_1, x2_1, x3_1, has_ff)
        # skip connection
        x1_ = Add()([x1_, x1_1])
        x2_ = Add()([x2_, x2_1])
        x3_ = Add()([x3_, x3_1])

    # set2set for both the atom and bond
    node_vec = Set2Set(T=n_pass, n_hidden=n3)([x1_, x6])
    edge_vec = Set2Set(T=n_pass, n_hidden=n3)([x2_, x7])
    # concatenate atom, bond, and global
    final_vec = Concatenate(axis=-1)([node_vec, edge_vec, x3_])
    if dropout:
        final_vec = Dropout(dropout)(final_vec)
    # final dense layers
    final_vec = Dense(n2, activation=act)(final_vec)
    final_vec = Dense(n3, activation=act)(final_vec)

    if is_classification:
        final_act = 'sigmoid'
        loss = 'binary_crossentropy'
    else:
        final_act = None
        loss = loss

    out = Dense(n_target, activation=final_act)(final_vec)
    model = KerasModel(inputs=[x1, x2, x3, x4, x5, x6, x7], outputs=out)
    model.compile(Adam(lr), loss)

    return GraphModel(model=model, graph_convertor=graph_convertor,
                      distance_convertor=distance_convertor)


def load_megnet_model(fname):
    """
    Customized load_model for MEGNet, which requires some custom objects.
    :param fname: HDF5 file containing model.
    :return: GraphModel
    """
    from keras.utils import get_custom_objects
    from keras.models import load_model

    custom_objs = get_custom_objects()
    custom_objs.update({'mean_squared_error_with_scale': mse_scale,
                        'softplus2': softplus2,
                        'MEGNet': MEGNet,
                        'Set2Set': Set2Set})
    return load_model(fname, custom_objects=custom_objs)

